### PR TITLE
feat: OAuth flow - Register/Login/Consent

### PR DESCRIPTION
**Pull request by Nesqwik**
Closes:  
- #370 
- #372
- #373 

## Description of the changes
The basic implementation of the OAuth flow. This include : 
- Register Flow
- Login Flow
- Consent Flow

This does NOT include :
- Any Design implementation
- Any client-side code
- Any scope validation for the API
- Any scope saved in database 
- Any backoffice UI to create a client

## Checklist
- [ ] I didn't over-scope my PR
- [ ] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I did not include breaking changes
- [ ] I made my own code-review before requesting one

### I included unit tests that cover my changes
  - [ ] 👍 yes
  - [ ] 🙅 no, because they aren't needed
  - [ ] 🙋 no, because I need help

### I added/updated the documentation about my changes

- [ ] 📜 README.md
- [ ] 📕 docs/*.md
- [ ] 📓 docs.lenra.io
- [ ] 🙅 no documentation needed

## Technical highlight/advice
How to test it : 
- Start dependancies (Hydra + Db) ```docker-compose up -d```
- Start server ```mix phx.server```
- Create the Hydra client using the CLI : 
```bash
>docker compose exec hydra hydra create client --endpoint http://127.0.0.1:4445 --token-endpoint-auth-method none --scope foo --redirect-uri https://oauthdebugger.com/debug
CLIENT ID       <random_uuid>
CLIENT SECRET
GRANT TYPES     authorization_code
RESPONSE TYPES  code
SCOPE           foo
AUDIENCE
REDIRECT URIS   https://oauthdebugger.com/debug
```
- Go to https://oauthdebugger.com
  - Authorize URI : http://localhost:4444/oauth2/auth
  - Client ID : the one above
  - scope : foo
  - [x] Use PKCE?
  - Token URI : http://localhost:4444/oauth2/token
  - Send Request !

### CODE REVIEW

File: apps/identity_web/lib/hydra.ex [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+  end
+
+  @doc """
+  Check whether an ORY Hydra service is alive.
+  """
+  @spec check_alive :: Operation.t()
+  def check_alive do
+    %Operation{
+      method: :get,
+      path: "/health/alive"
```
Comment (taorepoara) [inline apps/identity_web/lib/hydra.ex:62]: I think that only health is needed here

```suggestion
      path: "/health"
```

File: docker-compose.yml [hunk 2]
```diff
@@ -36,7 +36,7 @@ services:
     restart: on-failure
   postgres:
     container_name: lenra-postgres
-    image: postgres:13
+    image: postgres:12
```
Comment (taorepoara) [inline docker-compose.yml:39]: Why did you downgraded the Postgres version ? We use Postgres 13 every where (even in production):
```suggestion
    image: postgres:13
```

**Review by taorepoara — CHANGES_REQUESTED** [review]

File: hydra/hydra.yml [hunk 3]
```diff
@@ -5,6 +5,7 @@ serve:
     cors:
       enabled: true
       allowed_origins:
+        - https://oauthdebugger.com
         - http://localhost:10000
```
Comment (shiipou) [inline hydra/hydra.yml:9]: Can you use environment variables to setup this ? It's easier for defining values for our environments

Comment (Nesqwik) [reply hydra/hydra.yml:9]: It should not be used outside of development purpose.
I don't know if env variable are workig in this yaml with Hydra. 
I could just allow everything to simplify.

File: docker-compose.yml (hunk 2 above)
Comment (Nesqwik) [reply docker-compose.yml]: nvm it's working.

**Comment by taorepoara** [comment]
@Nesqwik The CI fails, you have to resolve it before a new revew

**Comment by Nesqwik** [comment]
@taorepoara Should be good :)

**Review by taorepoara — APPROVED** [review]

**Pull request merged**